tasks/update-granules-metadata/tests/test-index.js

```diff
@@ -0,0 +1,122 @@
+}
+
+async function uploadFiles(files, bucket) {
+  await Promise.all(files.map((file) => promiseS3Upload({
+    Bucket: bucket,
+    Key: parseS3Uri(file).Key,
+    Body: !(file.endsWith('.cmr.xml') || file.endsWith('.cmr.json'))
+      ? parseS3Uri(file).Key : cmrReadStream(file),
+  })));
+}
+
+function granulesToFileURIs(granules) {
+  const s3URIs = granules.reduce((arr, g) => arr.concat(g.files.map((file) => file.filename)), []);
+  return s3URIs;
+}
+
+function buildPayload(t) {
+  const newPayload = t.context.payload;
+
+  newPayload.config.bucket = t.context.stagingBucket;
+  newPayload.config.buckets.internal.name = t.context.stagingBucket;
+  newPayload.config.buckets.public.name = t.context.publicBucket;
+  newPayload.config.buckets.protected.name = t.context.protectedBucket;
+
+  newPayload.input.granules.forEach((gran) => {
+    gran.files.forEach((file) => {
+      file.bucket = t.context.stagingBucket;
+      file.filename = buildS3Uri(t.context.stagingBucket, parseS3Uri(file.filename).Key);
+    });
+  });
+
+  return newPayload;
+}
+
+test.beforeEach(async (t) => {
+  t.context.stagingBucket = randomId('staging');
+  t.context.publicBucket = randomId('public');
+  t.context.protectedBucket = randomId('protected');
+  t.context.systemBucket = randomId('system');
+  t.context.stackName = randomString();
+  await Promise.all([
+    s3().createBucket({ Bucket: t.context.stagingBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.publicBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.protectedBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.systemBucket }).promise(),
+  ]);
+  process.env.system_bucket = t.context.systemBucket;
+  process.env.stackName = t.context.stackName;
+  putJsonS3Object(
+    t.context.systemBucket,
+    getDistributionBucketMapKey(t.context.stackName),
+    {
+      [t.context.stagingBucket]: t.context.stagingBucket,
+      [t.context.publicBucket]: t.context.publicBucket,
+      [t.context.protectedBucket]: t.context.protectedBucket,
+      [t.context.systemBucket]: t.context.systemBucket,
+    }
+  );
+
+  const payloadPath = path.join(__dirname, 'data', 'payload.json');
+  const rawPayload = fs.readFileSync(payloadPath, 'utf8');
+  t.context.payload = JSON.parse(rawPayload);
+  const filesToUpload = granulesToFileURIs(t.context.payload.input.granules);
+  t.context.filesToUpload = filesToUpload.map((file) =>
+    buildS3Uri(`${t.context.stagingBucket}`, parseS3Uri(file).Key));
+  process.env.REINGEST_GRANULE = false;
+});
+
+test.afterEach.always(async (t) => {
+  await recursivelyDeleteS3Bucket(t.context.publicBucket);
+  await recursivelyDeleteS3Bucket(t.context.stagingBucket);
+  await recursivelyDeleteS3Bucket(t.context.protectedBucket);
+  await recursivelyDeleteS3Bucket(t.context.systemBucket);
+});
+
+test.serial('Should add etag to each CMR metadata file by checking that etag is one of more characters not whitespace', async (t) => {
```

Comment:
I think it might be helpful to have another test where the incoming payload already has an etag on the CMR file and verify that the etag changes after running the task. This scenario would not occur for any of current workflows, but it could if people use this task later in a workflow chain. What do you think?

Reply:
@markdboyd 
Yes, I think that is useful to test. I think I need more context on what that workflow could look like. Would a workflow chain in which that scenario might occur look like this:

```
      sync_granule_task_arn: module.cumulus.sync_granule_task.task_arn,
      add_missing_file_checksums_task_arn: 
      module.cumulus.add_missing_file_checksums_task.task_arn,
      fake_processing_task_arn: module.cumulus.fake_processing_task.task_arn,
      files_to_granules_task_arn: module.cumulus.files_to_granules_task.task_arn,
      move_granules_task_arn: module.cumulus.move_granules_task.task_arn,
      hyrax_metadata_updates_task_arn: module.cumulus.hyrax_metadata_updates_task.task_arn,
      post_to_cmr_task_arn: module.cumulus.post_to_cmr_task.task_arn,
      update_granules_metadata_task_arn: module.cumulus.update_granules_metadata_task.task_arn
```
where it updates the granules metadata last? Correct me if I am wrong, but etags are updated on any s3 PUT event, right? So, in this case, hyrax metadata update will write to s3 and it will consequently update the etag. Let me know if there is a fault in my understanding/reasoning!

Reply:
> etags are updated on any s3 PUT event

Yes, that is correct. 

That workflow you outlined would trigger the behavior - but I don't think it would ever occur in reality. Once `post_to_cmr ` is run, the metadata on S3 has already been posted to CMR , so there's no need to update the metadata filed stored on S3.

Really, I can't actually think of a real life workflow where update_granules_metadata might run and receive etags as part of the input, but keep in mind that users can assemble our tasks into workflows however they want including with their own custom tasks. So my motivation for testing this is because it is **theoretically** possible for the task to receive input with etags on the files and it seems easy enough to test.